File: 00_assignments/01a_file_formats/node/dataParser.js
const fs = require("fs");
const csv = require("csv-parser");
const yaml = require("yaml");
const xml2js = require("xml2js");

function parseCsv() {
    const results = [];
    fs.createReadStream("../me.csv")
        .pipe(csv())
        .on("data", (data) => results.push(data))
        .on("end", () => {
            console.log(results);
        });
}

function parseJson() {
    fs.readFile("../me.json", "utf8", (err, data) => {
        if (err) {
            console.log(err);
            return;
        }
        const parsedData = JSON.parse(data);
        console.log(parsedData);
    });
}

function parseYaml() {
    fs.readFile("../me.yaml", "utf8", (err, data) => {
        if (err) {
            console.log(err);
            return;
        }
        const parsedData = yaml.parse(data);
        console.log(parsedData);
    });
}

function parseXml() {
    const parser = new xml2js.Parser();
    fs.readFile("../me.xml", (err, data) => {
        if (err) {
            console.log(err);
            return;
        }
        parser.parseString(data, (err, result) => {
            const parsedData = result;
            console.log(parsedData);
        });
    });
}

function parseTxt() {
    fs.readFile("../me.txt", "utf8", (err, data) => {
        const lines = data.split("\n");

        const object = {};

        lines.forEach((l) => {
            // replace unrequired/filler characters and split into an array of key value pairs
            const replaceCharacters = l.replaceAll(":", "").replaceAll(",", "");
            const splitArray = replaceCharacters.split(" ");

            let key = splitArray[0];

            // remove first element of array since that is the key
            splitArray.shift();
            let value = splitArray;

            if (key == "name" || key == "age") value = value.toString();

            object[key] = value;
        });

        console.log(object);
    });
}

parseCsv();
parseJson();
parseYaml();
parseXml();
parseTxt();

File: 00_assignments/01a_file_formats/python/data_parser.py
import csv
import json
import yaml
import xml.etree.ElementTree as ET


def parse_txt():
    with open('../me.txt', 'r') as f:
        text = f.readlines()

        data = {}
        for line in text:
            key, value = line.strip().split(': ', 1)
            if ',' in value:
                value = [item.strip() for item in value.split(',')]
            data[key] = value

        print(data)


def parse_csv():
    with open('../me.csv', 'r') as f:
        csv_reader = csv.reader(f)
        for row in csv_reader:
            print(row)


def parse_json():
    with open('../me.json', 'r') as f:
        parsed_data = json.load(f)
        print(parsed_data)


def parse_yaml():
    with open('../me.yaml', 'r') as f:
        parsed_data = yaml.safe_load(f)
        print(parsed_data)


def parse_xml():
    root = ET.parse('../me.xml')
    parsed_dict = dict()
    for child in root.iter():
        if child.text.strip():
            parsed_dict[child.tag] = child.text
    print(parsed_dict)


parse_txt()
# parse_csv()
# parse_json()
# parse_yaml()
# parse_xml()
